dex/fiatrates/oracle.go

```diff
@@ -0,0 +1,274 @@
+	// discarded.
+	FiatRateDataExpiry = 60 * time.Minute
+
+	// averageRateRefreshInterval is how long it'll take before a fresh fiat
+	// average rate is calculated.
+	averageRateRefreshInterval = defaultRefreshInterval + time.Minute
+)
+
+// Oracle manages and retrieves fiat rate information from all enabled rate
+// sources.
+type Oracle struct {
+	log      dex.Logger
+	sources  []*source
+	ratesMtx sync.RWMutex
+	rates    map[string]*FiatRateInfo
+}
+
+func NewFiatOracle(cfg Config) (*Oracle, error) {
+	fiatOracle := &Oracle{
+		rates:   make(map[string]*FiatRateInfo),
+		sources: fiatSources(cfg),
+	}
+
+	assets := strings.Split(cfg.Assets, ",")
+	for _, asset := range assets {
+		_, ok := dex.BipSymbolID(strings.ToLower(asset))
+		if !ok {
+			return nil, fmt.Errorf("unknown asset %s", asset)
+		}
+
+		// Initialize entry for this asset.
+		fiatOracle.rates[parseSymbol(asset)] = new(FiatRateInfo)
+	}
+
+	return fiatOracle, nil
+}
+
+// Rate returns the current fiat rate information for the provided symbol.
+// Returns zero if there are no rates for the provided symbol yet.
+func (o *Oracle) Rate(ctx context.Context, symbol string) float64 {
+	o.ratesMtx.Lock()
+	defer o.ratesMtx.Unlock()
+
+	symbol = parseSymbol(symbol)
+	rateInfo := o.rates[symbol]
+	hasRateInfo := rateInfo != nil
+	if hasRateInfo && time.Since(rateInfo.LastUpdate) < FiatRateDataExpiry && rateInfo.Rate > 0 {
+		return rateInfo.Rate
+	}
+
+	if !hasRateInfo {
+		o.fetchFiatRateNow(ctx, symbol)
+	}
+
+	return 0
+}
+
+// fetchFiatRateNow attempts to retrieve fiat rate from enabled sources and
+// creates an entry for the provided asset. It'll try until one source returns a
+// non-zero value for the provided symbol. symbol must have been parsed by
+// parseSymbol function. Must be called with a write lock held on o.ratesMtx.
+func (o *Oracle) fetchFiatRateNow(ctx context.Context, symbol string) {
+	var fiatRate float64
+	defer func() {
+		// Initiate an entry for this asset. Even if fiatRate is still zero when
+		// we get here, data for this asset will be fetched in the next refresh
+		// cycle.
+		o.rates[symbol] = &FiatRateInfo{
+			Rate:       fiatRate,
+			LastUpdate: time.Now(),
+		}
+	}()
+
+	for _, s := range o.sources {
+		if s.isDisabled() {
+			continue
+		}
+
+		newRate, err := s.getRates(ctx, []string{symbol}, o.log)
+		if err != nil {
+			o.log.Errorf("%s.getRates error: %v", s.name, err)
+			continue
+		}
+
+		fiatRate = newRate[symbol]
+		if fiatRate > 0 {
+			break
+		}
+	}
+}
+
+// assets retrieves all assets that data can be fetched for.
+func (o *Oracle) assets() []string {
+	o.ratesMtx.RLock()
+	defer o.ratesMtx.RUnlock()
+	var assets []string
+	for sym := range o.rates {
+		assets = append(assets, sym)
+	}
+	return assets
+}
+
+// Run starts goroutines that refresh fiat rates every source.refreshInterval.
+// This should be called in a goroutine as it's blocking.
+func (o *Oracle) Run(ctx context.Context, log dex.Logger) {
+	o.log = log
```

Comment:
Can't pass the logger to the constructor?

Reply:
> Can't pass the logger to the constructor

We'd need that in two methods, would look a bit messy I think.

Reply:
Huh? Pass it to the constructor and assign it to a field like we everywhere. What am I missing?

Reply:
Dang, I think I missed your point a while back.